### PR TITLE
[RewardTrainer] Tokenize inputs within trainer

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR aligns the `RewardTrainer` with the other TRL trainer to apply tokenization within the trainer itself. This has the nice effect of simplifying the example script significantly.

The training logs before/after this PR look within noise from random seed IMO

Screenshot 2024-09-23 at 14 21 36

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/reward_trainer.py [hunk 1]
```diff
@@ -205,6 +225,33 @@ def __init__(
...
+        if "input_ids" not in train_dataset.column_names:
+            with PartialState().local_main_process_first():
+                fn_kwargs = {"tokenizer": tokenizer}
+                train_dataset = train_dataset.map(
+                    _tokenize,
+                    batched=True,
+                    fn_kwargs=fn_kwargs,
+                    num_proc=args.dataset_num_proc,
+                )
+                train_dataset = train_dataset.filter(
```
Comment (lewtun) [inline trl/trainer/reward_trainer.py:238]: This filter is important because otherwise you get samples that exceed the model's context length and get truncated => noisy signal because the model can miss the proper label

I could remove it and leave it up to the user's discretion if you prefer.

**Review by kashif — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -132,33 +129,7 @@ model = AutoModelForSequenceClassification.from_pretrained(
...
-    return new_examples
-
-chosen_fn = conversations_formatting_function(tokenizer, "chosen")
-rejected_fn = conversations_formatting_function(tokenizer, "rejected")
-dataset = dataset.map(lambda x: {"chosen": chosen_fn(x), "rejected": rejected_fn(x)})
-dataset = dataset.map(
-    preprocess_function,
-    batched=True,
-)
+dataset = dataset.map(maybe_apply_chat_template, fn_kwargs={"tokenizer": tokenizer})
```
Comment (qgallouedec) [inline README.md:132]: what about adding it in the trainer as well?

Comment (lewtun) [reply README.md:132]: Yes, I think this could be nice to make it consistent with the `SFTTrainer`! I'll push a change and fix the tests. 

We should later apply this to the other trainers

File: trl/trainer/reward_trainer.py (hunk 1 above)
Comment (qgallouedec) [inline trl/trainer/reward_trainer.py:238]:
```suggestion
                )
                # This filter is important because otherwise you get samples that exceed the model's context length and
                # get truncated => noisy signal because the model can miss the proper label. The downside is that the
                # user might get surprised if N samples are missing from training.
                train_dataset = train_dataset.filter(
```

**Comment by qgallouedec** [comment]
Can you add/modify the tests? You should be able to use `url-internal-testing/zen` as done in the other tests.

File: README.md (hunk 2 above)
Comment (qgallouedec) [reply README.md:132]: Yes! see #2071

Comment (lewtun) [reply README.md:132]: Done in https://github.com/huggingface/trl/pull/2102/commits/13b5ed0fafc3e218273ab96f530e3fc7a17453af

**Comment by lewtun** [comment]
@qgallouedec I've refactored the tests to mostly use the `zen` dataset - let me know if you want other parts tested or if this is good to merge

File: tests/test_reward_trainer.py [hunk 3]
```diff
@@ -14,159 +14,111 @@
...
+from trl.trainer.reward_trainer import _tokenize
+
+
+if is_peft_available():
+    from peft import LoraConfig, TaskType
 
 
 class RewardTrainerTester(unittest.TestCase):
+    def setUp(self):
+        self.model_id = "hf-internal-testing/tiny-random-LlamaForCausalLM"
```
Comment (lewtun) [inline tests/test_reward_trainer.py:34]: I think we should gradually move towards testing the most popular LLM architectures instead of relying on `gpt2` which has a bunch of annoying things like a missing PAD token

File: README.md [hunk 4]
```diff
@@ -120,7 +117,7 @@ trainer.train()
 Here is a basic example on how to use the `RewardTrainer`:
 
 ```python
-from trl import RewardConfig, RewardTrainer
+from trl import RewardConfig, RewardTrainer, maybe_apply_chat_template
```
Comment (qgallouedec) [inline README.md:120]:
```suggestion
from trl import RewardConfig, RewardTrainer
```

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
LGTM!

File: trl/trainer/reward_config.py [hunk 5]
```diff
@@ -36,8 +36,12 @@ class RewardConfig(TrainingArguments):
...
             https://huggingface.co/papers/2312.09244, Eq. 2). Recommended value: `0.01`.
+        remove_unused_columns (`bool`, *optional*, defaults to `False`):
+            Whether or not to remove the columns that are not used by the model's forward pass. Can be `True` only if
+            the dataset is pretokenized.
     """
 
     max_length: Optional[int] = None
     dataset_num_proc: Optional[int] = None
     center_rewards_coefficient: Optional[float] = None
+    remove_unused_columns: Optional[bool] = False
```
Comment (qgallouedec) [inline trl/trainer/reward_config.py:47]:
```suggestion
    remove_unused_columns: bool = False
```

**Review by kashif — APPROVED** [review]
